unblob/handlers/archive/ewf/ewf.py

```diff
@@ -0,0 +1,157 @@
+    VOLUME = b"volume" + b"\x00" * 10
+
+
+def find_chunk_size(header: Instance) -> int:
+    size = header.sectors_per_chunks
+    log = math.log(size, 2)
+    power = math.pow(2, log + 9)
+    return int(power)
+
+
+def is_valid_header(header: Instance) -> bool:
+    if header.field_start != 0x01 or header.field_end != 0x0:
+        return False
+    return True
+
+
+def calculate_crc(file: File, start_offset: int, size: int) -> int:
+    digest = 0
+    for chunk in iterate_file(file, start_offset, size):
+        digest = zlib.adler32(chunk, digest)
+    return (digest ^ -1) & 0xFFFFFFFF
+
+
+class EWFExtractor(Extractor):
+    def __init__(self, header_struct: str):
+        self.header_struct = header_struct
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        sectors_per_chunk = 0
+        volume_descriptor = None
+        outpath = outdir.joinpath(f"{inpath.name}.decrypted")
+        with File.from_path(inpath) as file:
+            file.seek(13)  # we skip the initial header
+            data_descriptor = self._struct_parser.parse(
+                "data_descriptor_t", file, Endian.LITTLE
+            )
+            logger.debug("data_descriptor_t", header=data_descriptor, _verbosity=3)
+            # the file is made of section, we loop over all the sections
+            while data_descriptor.definition != Definition.DONE.value:
+                file.seek(data_descriptor.next_offset, io.SEEK_SET)
+                data_descriptor = self._struct_parser.parse(
+                    "data_descriptor_t", file, Endian.LITTLE
+                )
+                logger.debug("data_descriptor_t", header=data_descriptor, _verbosity=3)
+
+                if data_descriptor.definition == Definition.VOLUME.value:
+                    volume_descriptor = self._struct_parser.parse(
+                        "volume_structure_t", file, Endian.LITTLE
+                    )
+                    logger.debug(
+                        "volume_descriptor", header=volume_descriptor, _verbosity=3
+                    )
+                    sectors_per_chunk = find_chunk_size(volume_descriptor)
+
+                if data_descriptor.definition == Definition.SECTOR.value:
+                    with outpath.open("a+b") as outfile:
+                        if volume_descriptor is not None:
+                            remaining_size = data_descriptor.section_size - len(
+                                data_descriptor
+                            )
+                            while remaining_size > 0:
+                                chunk_size = min(sectors_per_chunk, remaining_size)
+                                for chunk in iterate_file(
+                                    file, file.tell(), chunk_size
+                                ):
+                                    outfile.write(chunk[:-4])
+                                remaining_size -= chunk_size
```

Comment:
This is way too complex, I'm even amazed that flake did not report it. I would split the work in dedicated functions, depending on the definition type.

Reply:
Divided the code into different sections